File: 713. Subarray Product Less Than K/Solution.cs
public class Solution {
    public int NumSubarrayProductLessThanK(int[] nums, int k) {
        
        int numValidSubarrays = 0;

        for (int l = 0, r = 0, product = 1; r < nums.Length; ++r) {

            product *= nums[r];
            while (product >= k && l <= r) {
                product /= nums[l++];
            }
            numValidSubarrays += r-l+1;
        }

        return numValidSubarrays;
    }
}

File: 713. Subarray Product Less Than K/solution.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:

        l, numValidSubarrays, product = 0, 0, 1

        for r, num in enumerate(nums):

            product *= num
            while product >= k and l <= r:
                product /= nums[l]
                l += 1
            
            numValidSubarrays += r-l+1

        return numValidSubarrays
